servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/ByteArrayMapper.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright © 2024 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import io.servicetalk.concurrent.api.FromInputStreamPublisher.ToByteArrayMapper;
+
+import static io.servicetalk.concurrent.api.FromInputStreamPublisher.DEFAULT_MAX_BUFFER_SIZE;
+import static io.servicetalk.concurrent.api.FromInputStreamPublisher.ToByteArrayMapper.DEFAULT_TO_BYTE_ARRAY_MAPPER;
+
+/**
+ * A mapper to transform {@code byte[]} buffer regions into a desired type {@code T}.
+ *
+ * @param <T> Type of the result of this mapper
+ */
+@FunctionalInterface
+public interface ByteArrayMapper<T> {
+
+    /**
+     * Maps a specified {@code byte[]} buffer region into a {@code T}.
+     * <p>
+     * The mapper can operate only with the specified region of the {@code buffer}. Access to other parts of the buffer
+     * may lead to unexpected results. The specified region can be safely used by the mapper without a need to copy
+     * data.
```

Comment:
It says "Access to other parts of the buffer may lead to unexpected results.". Do you have a recommendation on how to rephrase it to highlight what you suggested?

> It's pretty common for people to wrap with a ByteBuffer and set limits, but still allow access to the underlying array.

It's possible to extract the underlying array from the `HeapByteBuffer`. However, users typically respect offset and length based on ByteBuffer info available via getters. Otherwise, it's always a bug.

Reply:
> It says "Access to other parts of the buffer may lead to unexpected results.". Do you have a recommendation on how to rephrase it to highlight what you suggested?

Maybe something to the tune of "Access to other parts of the buffer may lead to unexpected results and due care should be taken to avoid leaking that data through the result type T." ?

> It's possible to extract the underlying array from the HeapByteBuffer. However, users typically respect offset and length based on ByteBuffer info available via getters. Otherwise, it's always a bug.

The array can't be accessed if the HeapByteBuffer is read-only, but I agree that it's uncommon to do that. This was just an example to justify making the warning a bit more explicit that people should be careful to avoid leaking the buffer through their type T.